### ISSUE TITLE
CycleGAN Fake B looks very similar to real A (vice versa)

### ISSUE THREAD

**Issue by momenator** [junyanz/pytorch-CycleGAN-and-pix2pix#756]
Hi,

I've trained a CycleGAN model to do unpaired translation of medical imaging of domain A to domain B (and vice versa). My input data are 3d volumes of the human body, which are transformed into 2d 256x256 (only 1 channel) patches so that they can be passed into the network for training. After preprocessing, I have 160k patches of domain A and 59k of patches of domain B. 

For training, I used the following flags: 

`--model cycle_gan --gpu_ids 6,3,0,5 --preprocess none --dataset_mode unaligned --input_nc 1 --output_nc 1 --max_dataset_size 55000 --num_threads 24 --batch_size 24`

I let it run for 50 epochs and the result is that real A and fake B looks very similar (and so is real B and fake A).

This is the result after 50th epoch during training: 
<img width="530" alt="Screenshot 2019-09-05 at 10 56 56" src="https://user-images.githubusercontent.com/6440096/64327507-1db21980-cfcc-11e9-8dab-31068495ffed.png">

This is the result on the test set. As you can see, fake B is still in the same domain as that of real A.
<img width="257" alt="Screenshot 2019-09-05 at 11 01 36" src="https://user-images.githubusercontent.com/6440096/64327794-93b68080-cfcc-11e9-94c2-e5406144d760.png">

It seems like the network is overfitting, but I also thought it might be trying to do something else (I'm not so sure). 

Does anyone have any idea of what went wrong here and the possible remedies? (apart from data augmentation, which I will try out next). Any help or feedback is greatly appreciated.

**Comment by junyanz** [comment]
Could you share with us images for `real_B`? If you want to have bigger changes, you can try to set `--lambda_identity 0`.

**Comment by momenator** [comment]
As a reference real_A looks like the following:

![Screenshot 2019-09-05 at 17 03 02](https://user-images.githubusercontent.com/6440096/64354246-57e8de80-cfff-11e9-8909-2348723ed1a2.png)

real_B:

![Screenshot 2019-09-05 at 17 04 01](https://user-images.githubusercontent.com/6440096/64354264-63d4a080-cfff-11e9-88ff-c667b9ee1ba9.png)

Those two examples are slices from the 3d volume, which are then processed into 256x256 image patches.

**Comment by junyanz** [comment]
256x256 patches might be too small compared to the original images. You also may want to crop the center region from real_B. It's hard to synthesize head and limbs from real_A.

**Comment by ibro45** [comment]
@momenator it's been a while, but could you tell if you were able to solve this issue and how? Thanks!

**Comment by momenator** [comment]
@ibro45 sorry for the late reply. I got it to work using the following approaches:
1. Crop the MR scans such that it covers the same area as that of CT (the torso)
2. Make sure to centre the data (e.g x - x.mean) over the current volume of scan, followed by a normalisation of your choice (e.g division by x.std() should be fine). Before I used (x - x.min) / (x.max - x.min) and this did not work very well.
3. The images are too big, so I used a patch-based approach using 256x256 patches and then reconstruct the big scan accordingly. 

The implementation is here: https://github.com/momenator/CycleGAN if you are interested.
